Initial commit -- Adding calibration loss specific to segmentation

Bala93 opened the pull request:
### Description

Model calibration has helped in developing reliable deep learning models. In this pull request, I have added a new loss function NACL (https://arxiv.org/abs/2303.06268, https://arxiv.org/abs/2401.14487) which has shown promising results for both discriminative and calibration in segmentation. 

**Future Plans:** Currently, MONAI has some of the alternative loss functions (Label Smoothing, and Focal Loss), but it doesn't have the calibration specific loss functions (https://arxiv.org/abs/2111.15430, https://arxiv.org/abs/2209.09641). Besides, these methods are better evaluated with calibration metrics, Expected Calibration Error (https://lightning.ai/docs/torchmetrics/stable/classification/calibration_error.html).

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

ericspod commented on monai/losses/segcalib.py:
```diff
@@ -0,0 +1,173 @@
+
+class NACLLoss(_Loss):
+    """
+    Murugesan, Balamurali, et al.
```
Please add here a brief description of what the loss function does, what applications it's for, and a bit more description of what the arguments do and mean.

ericspod commented on monai/losses/segcalib.py:
```diff
@@ -0,0 +1,173 @@
+
+        utargets = self.get_constr_target(targets)
+
+        if self.distance_type == "l1":
+            loss_conf = torch.abs(utargets - inputs).mean()
+
+        if self.distance_type == "l2":
+            loss_conf = (torch.abs(utargets - inputs) ** 2).mean()
```
```suggestion
        if self.distance_type == "l1":
            loss_conf = utargets.sub(inputs).abs_().mean()
        elif self.distance_type == "l2":
            loss_conf = utargets.sub(inputs).pow_(2).abs_().mean()
```
Maybe a little faster?

ericspod commented on monai/losses/segcalib.py:
```diff
@@ -0,0 +1,173 @@
+
+        loss = loss_ce + self.alpha * loss_conf
+
+        return loss  # , loss_ce, loss_conf
```
```suggestion
        return loss 
```

ericspod reviewed (commented):
Thanks for the contribution! I had a few comments, and the CICD errors are easy fixes I think. The test fail is from accuracy it seems which can be fixed by changing the error.

Bala93 commented on monai/losses/segcalib.py:
```diff
@@ -0,0 +1,173 @@
+    return gaussian_kernel / torch.sum(gaussian_kernel)
+
+
+class GaussianFilter(torch.nn.Module):
```
Hi @ericspod, I have integrated the support for both 2d and 3d now. I can migrate to GaussianFilter and MeanFilter in some time. Thanks.

ericspod replied:
If you can move to our existing definitions with the same results that would be great, thanks.

ericspod commented on monai/losses/segcalib.py:
I would rename this file to `nacl_loss.py` to match the naming of other loss files in this directory.

ericspod commented on tests/test_nacl_loss.py:
```diff
@@ -0,0 +1,228 @@
+        },
+        1.1850,
+    ],
+    [
+        {"classes": 3, "dim": 2, "kernel_ops": "gaussian"},
+        {
+            "inputs": torch.tensor(
+                [
+                    [
+                        [
+                            [0.0411, 0.3386, 0.8352, 0.2741],
+                            [0.4821, 0.0519, 0.2561, 0.9391],
+                            [0.5954, 0.4184, 0.9160, 0.7977],
+                            [0.0588, 0.9156, 0.1307, 0.9914],
+                        ],
+                        [
+                            [0.8481, 0.0892, 0.2044, 0.8470],
+                            [0.3558, 0.3569, 0.0979, 0.4491],
+                            [0.0876, 0.0929, 0.4040, 0.8384],
+                            [0.5313, 0.3927, 0.4165, 0.1107],
+                        ],
+                        [
+                            [0.7993, 0.6938, 0.3151, 0.8728],
+                            [0.7332, 0.4111, 0.3862, 0.9988],
+                            [0.2622, 0.5002, 0.1905, 0.1644],
+                            [0.6354, 0.0047, 0.1649, 0.7112],
+                        ],
+                    ]
+                ]
+            ),
+            "targets": torch.tensor([[[1, 2, 0, 1], [0, 2, 1, 2], [0, 0, 2, 1], [1, 1, 1, 2]]]),
+        },
+        1.0375,
+    ],
```
This test appears to test the same thing as the previous? Does this add anything?

Bala93 replied:
Sure, got it. I will remove one of the case, it is just different values for the input and target tensors.

ericspod commented on tests/test_nacl_loss.py:
```diff
@@ -0,0 +1,228 @@
+                        ],
+                    ]
+                ]
+            ),
```
This tensor is the same as that in the previous test, we can save some lines by putting this into a temporary variable and used it here and in the previous test.

Bala93 replied:
Fixed this one.

ericspod commented:
Hi @Bala93 I have a few further comments, the fixes for the `mypy` issues should be minor and the DCO page tells you how to address that. Thanks!

Bala93 commented:
> Hi @Bala93 I have a few further comments, the fixes for the `mypy` issues should be minor and the DCO page tells you how to address that. Thanks!

Hi @ericspod, Thanks for your feedback. I have addressed all the issues. I will work on using the monai functions (for the gaussian, and mean) in some time. Until then, is it possible to merge it with the main branch and make the loss function available ? Thank you.

KumoLiu commented on docs/source/losses.rst:
```diff
@@ -93,6 +93,11 @@ Segmentation Losses
 .. autoclass:: SoftDiceclDiceLoss
     :members:
 
+`NACLLoss`
+~~~~~~~~~~~~~~~~~~~~
```
```suggestion
`NACLLoss`
~~~~~~~~~~
```

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+from monai.utils import pytorch_after
+
+
+def get_mean_kernel_2d(ksize: int = 3) -> torch.Tensor:
+    mean_kernel = torch.ones([ksize, ksize]) / (ksize**2)
+    return mean_kernel
+
+
+def get_mean_kernel_3d(ksize: int = 3) -> torch.Tensor:
```
Perhaps we can combine these two to one function and move it under this [folder](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/convutils.py) to make it to be a more common used utils. What do you think?

cc @ericspod @Nic-Ma

ericspod replied:
```python
def get_mean_kernel(ksize: int = 3, dim: int = 2) -> torch.Tensor:
    return torch.ones([ksize] * dim) / (ksize**dim)
```
Something like this.

Bala93 replied:
Sure, got the idea. Will work on this as well as for the gaussian (https://github.com/Project-MONAI/MONAI/pull/7819#discussion_r1705520061)

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+    return mean_kernel
+
+
+def get_gaussian_kernel_2d(ksize: int = 3, sigma: float = 1.0) -> torch.Tensor:
```
Also for these two gaussian kernel functions.

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+    return gaussian_kernel
+
+
+class GaussianFilter(torch.nn.Module):
```
As @ericspod suggested, can we integrate these class into the existing one?

Bala93 replied:
I have a doubt with GaussianFilter implementation. In the MONAI implementation, we have sigma and truncate. Is it possible for me to use the kernel size and sigma to decide the truncate value ? https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L554 This might not have much impact with the objective function, but would be good adding it.

KumoLiu replied:
The truncation range typically determines the kernel size; the kernel size must be large enough to accommodate the majority of the Gaussian distribution's energy. Of course, you can also extend the current function; just be careful to ensure compatibility. Thanks!

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+        return svls_normalized
+
+
+class MeanFilter(torch.nn.Module):
```
And for this class, would it be better to move it under monai/networks/layers/simplelayers.py?

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+
+        self.old_pt_ver = not pytorch_after(1, 10)
+
+    # def ce(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
please remove the comment lines here.

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,279 @@
+
+    #     return self.cross_entropy(input, target)  # type: ignore[no-any-return]
+
+    def get_constr_target(self, mask: torch.Tensor) -> torch.Tensor:
```
Could you please add a small docstring here to demenstrate what's this function for?

KumoLiu commented on tests/test_nacl_loss.py:
```diff
@@ -0,0 +1,163 @@
+targets = torch.tensor([[[1, 1, 1, 1], [1, 1, 1, 0], [0, 0, 0, 0], [0, 0, 0, 0]]])
+
+TEST_CASES = [
+    [{"classes": 3, "dim": 2}, {"inputs": inputs, "targets": targets}, 1.1820],
```
Could you please include more tests for other supported kernel_ops and arguments?

KumoLiu reviewed (commented):
Thanks for the contribution! Overall looks good to me. Several minor inline comments.

Bala93 commented:
Hi @KumoLiu @ericspod. I have addressed the comments with the latest commit. Thank you.

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,113 @@
+            raise ValueError("Kernel ops must be either mean or gaussian")
+
+        if dim not in [2, 3]:
+            raise ValueError("Supoorts 2d and 3d")
```
```suggestion
            raise ValueError(f"Support 2d and 3d, got dim={dim}.")
```

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,113 @@
+            raise ValueError("Supoorts 2d and 3d")
+
+        if distance_type not in ["l1", "l2"]:
+            raise ValueError("Distance type must be either L1 or L2")
```
```suggestion
            raise ValueError(f"Distance type must be either L1 or L2, got {distance_type}")
```

KumoLiu commented on monai/losses/nacl_loss.py:
```diff
@@ -0,0 +1,113 @@
+        rmask: torch.Tensor
+
+        if self.dim == 2:
+            oh_labels = F.one_hot(mask.to(torch.int64), num_classes=self.nc).contiguous().permute(0, 3, 1, 2).float()
```
Could you please add more description for the input shape for this function and also for the forward, such as the definition of dimension (we often assume all non-transform components assume input data shape [batch x channel x spatial_dims]) and also need to be one-hot or not. 

Here is an example in the `DiceCELoss`.
https://github.com/Project-MONAI/MONAI/blob/6c23fd06fc11667beedd0ba730d4104076a8db2d/monai/losses/dice.py#L776-L784

Bala93 replied:
In my current implementation, I am treating it like typical Cross entropy loss in Pytorch. Ex: Predictions will be BNHW(D) and labels will be BHW(D). B - Batch size, N - No. of classes, HW(D) - Image dimensions. Can I stick with this and add this in docstring ?

KumoLiu replied:
Sure, it's just align with monai's assumption.

Bala93 replied:
Have updated the docstring in both the functions following the example.

KumoLiu reviewed (commented):
Thanks for the quick update, few further comments inline.

ericspod commented:
Looks good now to me with the last minor comments addressed.

KumoLiu reviewed (approved):
Thanks for the quick update, looks good to me now. Will trigger more tests and try to merge this one.

KumoLiu commented:
/build
